code/modules/mob/living/bot/bot.dm

```diff
@@ -441,5 +475,84 @@
 
 			//if((dir & SOUTH) && (D.dir & (EAST|WEST)))		return !D.check_access(ID)
 			//if((dir & EAST ) && (D.dir & (NORTH|SOUTH)))	return !D.check_access(ID)
-		else return !D.check_access(ID)	// it's a real, air blocking door
+		else 
+			if(istype(D, /obj/machinery/door/airlock/lift))	//Stop bots committing suicide down the lift shaft, as amusing as it is
+				return !D.check_access(ID) || D.density
+			var/obj/machinery/door/airlock/A = D
+			if(istype(A) && (A.locked || A.isAllPowerLoss() || A.welded))	//Avoid airlocks that are bolted, welded, or have no power
+				return 1
+			return (!D.check_access(ID))	// it's a real, air blocking door
+
+	for(var/obj/structure/railing/D in loc)
+		if(!D.density)			continue
+		if(dir & D.dir)			return 1
 	return 0
+
+
+//Multi-Z AStar Procs
+
+/turf/proc/Euclidean3dDistance(turf/t)
+	var/euclid_dist = sqrt(Square(src.x - t.x) + Square(src.y - t.y) + Square(src.z - t.z))
+	var/currentPathweight = ((locate(/obj/machinery/door) in src) && pathweight == 1) ? 2 : pathweight
+	var/targetPathweight = ((locate(/obj/machinery/door) in t) && t.pathweight == 1) ? 2 : t.pathweight
```

Comment:
Might be nicer to add a flat 'door traversal cost' if there's a door in the turf rather than using a ternary expression.
It's easier to think about and tune up and it will work with turfs with higher pathweights.

For example, imagine a 'sticky floor' turf that slows down robots massively (pathweight e.g. 5) - our preferences in terms of speed would be: Normal > Normal+Door > Sticky > Sticky+Door. Using current logic, Sticky == Sticky+Door.